### DISCUSSION TITLE
Fediverse Integration: login with oneechanhax@mastodon.social and federate with other authentication servers.

### DISCUSSION THREAD

**Discussion by oneechanhax** [yushijinhun/authlib-injector#253] [Ideas]
It would be awesome to have servers work like the Mastedon product, or other Fediverse products like Peertube. Your login would be similar to email usage, as "username@domain.tld". 
https://fediverse.party/

XMPP already uses this and it works by verifying the SSL of the domain name, doing client-to-server connections, and more importantly the server-to-server connections to pass around messages between users and between group chats at domains.
https://en.wikipedia.org/wiki/JID_(Jabber)
https://xmpp.org/extensions/xep-0029.html 

The "Mojang Namespace" seems like a related feature already, 
```
-Dauthlibinjector.mojangNamespace={default|enabled|disabled} 
```

already having users named as username@mojang is cool, it provides an idea of how to do it already.  
under federation naming, the username formats would be "username@mojang.com"  or "username@minecraft.net" 

To perform this though, we also need a way to allow-list and deny-list specific domains since people can abuse the existence of federation. This isn't a downside or a bug, its a feature!  

Building a web of trusted authentication servers is a big task to ask of a community, but it can make this authlib-injector last a very long time in our hearts.   
I'm interested in hearing what other people think of this.  Thank you for hearing me out with this proposal ❤️ 

[Maybe we can implement plugins to make specific domains read-only to worlds and the like, to prevent griefing. Would make it easy to add players to your world, but would also need more heavy modifications than just authentication edits.]

**Comment by lamadaemon** [comment]
How are you going to manage player profiles (name, uuid, and skins) and authentication process when joining a server? 

Those servers will need a support for Yggdrasil Protocol. I think the best way is to modify Mastedon server for supporting Yggdrasil Protocols instead of changing in authlib-injector.

**Reply by oneechanhax** [reply]
The reply thing is a little weird, 
I written this in response. 
https://github.com/yushijinhun/authlib-injector/discussions/253#discussioncomment-9744987

**Comment by oneechanhax** [comment]
In response to: https://github.com/yushijinhun/authlib-injector/discussions/253#discussioncomment-9744843
Probably wont be able to drop-in replacement the mastedon stuff. 

If we do it like XMPP, you need 2(1) servers and a client.
First a user needs to tell the auth domain that it wishes to identify with, lets say gamersnexus.com, the username will be ricky(@gamersnexus.com) to login as. 
It needs to authenticate with the gamersnexus.com server first and it will receive its full username (ricky@gamersnexus.com).  

Second, they want to connect to your minecraft server spelunkingminers.net, the server will then present the authentication servers that It is using, but in this case they use a separate server to do authentication "linustechtips.com", 
in order to connect ricky@gamersnexus.com to spelunkingminers.net (using linustechtips.com to auth), the authentication servers themselves have to verify the SSL is correct (important as this proves identity and ownership)
 gamersnexus.com and linustechtips.com then do a little dance to be like, "yes this is the user in question" "hmm yes, allow them i shall"  
 
  linustechtips.com will tell the minecraft server that the user wishes to authenticate and has correctly identified themself, 
 gamersnexus.com will then reply with the response it had received to the client and the client will proceed to enter, 
 Now 
 ```
 ricky@gamersnexus.com has joined the spelunkingminers.net
 ```
 is in chat :) 
 
 
For user management and storage, I believe you can use random UIDs or using partially of the less significant bits (im unsure which those are) but there may be challenges at scale. 
This certainly creates a large exponential issue since it goes from 1 server managing its own auth to managing everything possible.  Random UID per authentication may be the best we have until a better method comes around. 
Each server creating the UIDs as the successful authentications arrive, and only after a successful authentication. 
 ( we should have a hard limit per domain for user creating authentications to prevent abuse, rate limiting is a must)

So in the case above after they have successfully authenticated server-server, linustechtips.com would have to create the definitive UID since it is authenticating the server being reached, and it would need to forward that information to both the server and the client.

**Reply by lamadaemon** [reply]
That's a good idea, but still not really necessary to modify authlib-injector. All you need to do is implementing a linking server for Mastedon. Players will use `<name>@mastedon.example.com` as their username to login, and the linking server will manage their profile (name, uuid, and skins).

 However there is currently no way to solve two players logged with different services that have the same name. This issue is somehow related to project like MultiLogin or MultiYggdrasil.

**Reply by lamadaemon** [reply]
What authlib-injector really need is like what you said, the namespaces. Not just Mojang namespaces but across servers. However this will require authlib-injector to support multiple Yggdrasil services at once which is hard to implement. Maybe you can open an issue on requesting a better namespace system.

**Reply by oneechanhax** [reply]
Not sure I want to use mastedon as a requirement to do this sort of domain namespacing. Seems quite, heavy. it may do the authentication using the ActivityPub protocol which we could build on top of. could maybe be used as a secondary o-auth login or something like that.
https://en.wikipedia.org/wiki/ActivityPub

I can code in golang pretty well so far as a beginner in go, (as far as singlefile projects go)
managing state, especially fedi-authentication state, by hand, seems like a difficult challenge and nightmare.

May be a good idea to fork one of the golang auth servers and can try patching it together in a branch somehow,
I cant promise anything, but I'll think about this since its something I really want to see. Idk much about the protocols but only one way to find out what goes wrong is to trial and error and document it. 

I would want to bring it up with them first to see what they think first and see if they can do it for me hehe.

**Reply by oneechanhax** [reply]
From first glance,  
The MultiLogin/MultiYggdrasil is pretty neat, the MultiLogin looks like its trying to do what I am trying to achieve but are limited somewhere in how the yggdrasil protocal works. 
 
I think instead of doing it minecraft server sided with multilogin or via the multiyggdrasil auth-injector fork, the authentication servers themselves need to do communication between themselves more. 

Auth server fork to handle this auth server-to-server communication may be a todo.

**Comment by evan-goode** [comment]
I've been thinking about this lately. What about restricting player UUIDs to be a 128-bit hash of the namespaced player name? i.e where drasl.unmojang.org is the authserver:

```
echo -n evan@drasl.unmojang.org | b3sum -l 16
1cac2200328b967a69511aa97b6c1fdd
```

the Minecraft server would verify that the UUID is correct after `POST https://unmojang.org/session/minecraft/hasJoined`. That way there would be no ambiguity of "what nameserver owns what UUID".

A 128-bit hash means only around $$2^{64}$$ work is needed to generate two names that collide, but I think this is not a big security problem. An attacker would still need $$2^{128}$$ work (infeasible) to brute-force a specific player name, or $$2^{128}/n$$ work to brute-force any one of $$n$$ player names.

It would be nice if clients could fetch a player's skin directly from the player's homeserver, but that's probably a bad idea since it would reveal client IP addresses.

Routes like `GET https://sessionserver.mojang.com/session/minecraft/profile/<uuid>` would still be a challenge since you wouldn't know which authserver to ask for the profile. Possible solutions might be clients or authservers keeping a cache of known player@server -> UUID mappings or modifying the client to always look up skins by player@server rather than UUID, like CustomSkinLoader. Not sure whether that would work in all cases, e.g. player heads.

**Reply by evan-goode** [reply]
This would not allow players to change their player name, however.

Another idea is to use a [distributed hash table](https://en.wikipedia.org/wiki/Distributed_hash_table) to map UUIDs to homeservers. The DHT could look like:

```
6ca297c2-2bf2-e3ae-4c4f-04f6811da2c4 -> 69627b62a12c493d82983c2435add144drasl.unmojang.org
862389ce-6b3e-9b82-1405-467f1c119694 -> b06797e8501447e1a84b5d04eb05d846foo.example.com
fc6d0654-6065-ff61-4294-005db7bdfbf7 -> 78feb0b0a2ad4cb9904f0fe9b1026a81littleskin.cn
```

Each value (right-hand side) is 128 random bits (a nonce, essentially) followed by the homeserver domain. Each key (left-hand side) is the 128-bit blake3 hash of the corresponding value, and represents a player UUID.

To resolve the UUID `6ca297c2-2bf2-e3ae-4c4f-04f6811da2c4` to a player name, you index the DHT to get `69627b62a12c493d82983c2435add144drasl.unmojang.org`, then verify that the hash of `69627b62a12c493d82983c2435add144drasl.unmojang.org` matches the UUID, then query `drasl.unmojang.org` to get the player name.

If only player UUIDs were longer, say 256 bits, then we could use the first half for identifying a player on a single homeserver and the last half as a hash of the homeserver domain to establish UUID ownership. Then each homeserver would need a smaller map of "homeserver domain hash" -> homeserver rather than the larger "player UUID -> player@homeserver" map. But with only 128 bits, I don't think that's enough space to both (1) uniquely identify a player on a given nameserver and (2) provide enough protection against brute-forcing a domain name to collide with some homeserver domain hash.
